tuna/miopen/celery_tuning/tuning.py

```diff
@@ -0,0 +1,377 @@
+
+  return worker_type, kwargs, fdb_attr, q_name
+
+
+#pylint: disable=too-many-locals
+def tune(library, job_batch_size=1000):
+  """tuning loop to spin out celery tasks"""
+
+  tune_start_t = time.time()
+  worker_type, kwargs, fdb_attr, q_name = prep_tuning(library)
+  res_set = ResultSet([])
+  job_start_t = None
+  job_end_t = None
+
+  while True:
+    try:
+      job_list = []
+      with DbSession() as session:
+        #get all the jobs from mySQL
+        job_start_t = time.time()
+        job_list = library.get_jobs(session, library.fetch_state,
+                                    library.set_state, library.args.session_id,
+                                    job_batch_size)
+
+        for i in range(0, len(job_list), job_batch_size):
+          batch_jobs = job_list[i:min(i + job_batch_size, len(job_list))]
+          if library.args.fin_steps:
+            entries = library.compose_work_objs_fin(session, batch_jobs,
+                                                    library.dbt)
+
+          entries = library.compose_work_objs_fin(session, job_list,
+                                                  library.dbt)
+          serialized_jobs = serialize_chunk(entries)
+
+          #build context for each celery task
+          for job, config in serialized_jobs:
+            context = {
+                'job': job,
+                'config': config,
+                'worker_type': worker_type,
+                'arch': library.dbt.session.arch,
+                'num_cu': library.dbt.session.num_cu,
+                'kwargs': kwargs,
+                'fdb_attr': fdb_attr
+            }
+
+            #calling celery task, enqueuing to celery queue
+            res_set.add(hardware_pick.apply_async((context,), queue=q_name))
+
+      if not job_list:
+        if not res_set:
+          return False
+        LOGGER.info('All tasks added to queue')
+        job_end_t = time.time()
+        break
+    except KeyboardInterrupt:
+      LOGGER.error('Keyboard interrupt caught, draining results queue')
+      results_gather(res_set, worker_type)
+
+  results_gather(res_set, worker_type)
+  tune_end_t = time.time()
+  LOGGER.info(
+      'Took {%.4f} min to tune, {%.4f} of which where used to enqueue jobs',
+      (tune_end_t - tune_start_t) % 60, (job_end_t - job_start_t) % 60)
+
+  return True
+
+
+def results_gather(res_set, worker_type):
+  """Start subproc to drain result queue and populate mysql DB"""
+  LOGGER.info('Started drain process')
+  #start draining result_queue in subprocess
+  drain_process = Process(target=drain_queue, args=[worker_type, DBT])
+  drain_process.start()
+
+  LOGGER.info('Gathering async results in callback function, blocking')
+  #waiting on all results to come in
+  _ = res_set.join(callback=result_callback)
+
+  #terminate result_queue drain process with special queue token (NONE,NONE)
+  LOGGER.info('Adding terminatig token to result_queue')
+  result_queue.put([None, None])
```

Comment:
launch separate result queue drain that writes back to mySQL, blocking res_set join that waits for all enqueued jobs to terminate